crates/holochain/src/sweettest/sweet_topos/fact/density.rs

```diff
@@ -0,0 +1,145 @@
+    /// Assumes that the graph is simple.
+    /// Petgraph DOES NOT make any of these assumptions, so we have to do this
+    /// ourselves.
+    pub fn max_edge_count(graph: &NetworkTopology) -> usize {
+        let node_count = graph.node_count();
+        node_count * (node_count - 1)
+    }
+
+    /// This is the number of edges that we want to have in the graph. It is
+    /// a truncation of the density times the maximum number of edges, so we're
+    /// rounding down.
+    pub fn target_edge_count(&self, graph: &NetworkTopology) -> usize {
+        (self.density * Self::max_edge_count(&graph) as f64) as usize
+    }
+}
+
+impl<'a> Fact<'a, NetworkTopology> for DenseNetworkFact {
+    fn mutate(
+        &mut self,
+        g: &mut Generator<'a>,
+        mut graph: NetworkTopology,
+    ) -> Mutation<NetworkTopology> {
+        let target_edge_count = self.target_edge_count(&graph);
+        let mut rng: _ = super::rng_from_generator(g);
+
+        // Add edges until we reach the desired density.
+        while graph.edge_count() < target_edge_count {
+            graph.add_random_simple_edge(&mut rng)?;
+        }
+
+        // Remove edges until we reach the desired density.
+        while graph.edge_count() > target_edge_count {
+            graph.remove_random_edge(&mut rng)?;
+        }
+
+        Ok(graph)
+    }
+
+    fn label(&self) -> String {
+        format!("DenseNetworkFact {{ density: {} }}", self.density)
+    }
+
+    fn labeled(self, _: impl ToString) -> Self {
+        todo!()
+    }
+}
+
+#[cfg(test)]
+pub mod test {
+    use super::DenseNetworkFact;
+    use crate::prelude::unstructured_noise;
+    use crate::sweettest::fact::partition::StrictlyPartitionedNetworkFact;
+    use crate::sweettest::fact::size::SizedNetworkFact;
+    use crate::sweettest::sweet_topos::network::NetworkTopology;
+    use contrafact::Fact;
+    use petgraph::dot::{Config, Dot};
+
+    /// Test that we can build a dense network fact with `DenseNetworkFact::new`.
+    #[test]
+    fn test_dense_network_fact_new() {
+        let a = DenseNetworkFact::new(0.5);
+        let b = DenseNetworkFact { density: 0.5 };
+        assert_eq!(a, b);
+    }
+
+    #[test]
+    fn test_sweet_topos_dense_network() {
+        let mut g = unstructured_noise().into();
+        let mut size_fact = SizedNetworkFact {
+            nodes: 12,
+            agents: 1..=2,
+        };
+        let mut density_fact = DenseNetworkFact { density: 0.3 };
+        let mut graph = NetworkTopology::default();
+        graph = size_fact.mutate(&mut g, graph).unwrap();
+        println!(
+            "{:?}",
+            Dot::with_config(
+                graph.as_ref(),
+                &[
+                    //     Config::GraphContentOnly,
+                    Config::NodeNoLabel,
+                    Config::EdgeNoLabel,
+                ],
+            )
+        );
+        graph = density_fact.mutate(&mut g, graph).unwrap();
+        println!(
```

Comment:
Prefer tracing in tests because it can be turned off more selectively than println?